tools/bro-to-vast/bro-to-vast.cpp

```diff
@@ -0,0 +1,304 @@
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+
+#include <caf/config_option_adder.hpp>
+
+#include <broker/bro.hh>
+#include <broker/broker.hh>
+
+#include <vast/data.hpp>
+#include <vast/defaults.hpp>
+#include <vast/error.hpp>
+#include <vast/event.hpp>
+#include <vast/expression.hpp>
+#include <vast/uuid.hpp>
+
+#include <vast/concept/parseable/parse.hpp>
+#include <vast/concept/parseable/vast/expression.hpp>
+#include <vast/concept/parseable/vast/uuid.hpp>
+
+#include <vast/system/writer_command_base.hpp>
+#include <vast/system/sink.hpp>
+
+#include <vast/detail/add_error_categories.hpp>
+#include <vast/detail/add_message_types.hpp>
+#include <vast/detail/assert.hpp>
+#include <vast/detail/overload.hpp>
+
+namespace {
+
+constexpr char control_topic[] = "/vast/control";
+constexpr char data_topic[] = "/vast/data";
+
+constexpr char default_address[] = "localhost";
+constexpr uint16_t default_port = 43000;
+
+// Our custom configuration with extra command line options for this tool.
+class config : public broker::configuration {
+public:
+  config() {
+    // As a stand-alone application, we reuse the global option group from CAF
+    // to avoid unnecessary prefixing.
+    opt_group{custom_options_, "global"}
+      .add<uint16_t>("port,p", "the port to listen at or connect to");
+  }
+};
+
+/// Converts VAST data to the corresponding Broker type.
+broker::data to_broker(const vast::data& data) {
+  return caf::visit(vast::detail::overload(
+    [](const auto& x) -> broker::data {
+      return x;
+    },
+    [](caf::none_t) -> broker::data {
+      return {};
+    },
+    [](const vast::pattern& x) -> broker::data {
+      return x.string();
+    },
+    [](const vast::address& x) -> broker::data {
+      auto bytes = reinterpret_cast<const uint32_t*>(x.data().data());
+      return broker::address{bytes, broker::address::family::ipv6,
+                             broker::address::byte_order::network};
+    },
+    [](const vast::subnet& x) -> broker::data {
+      auto bytes = reinterpret_cast<const uint32_t*>(x.network().data().data());
+      auto addr = broker::address{bytes, broker::address::family::ipv6,
+                                  broker::address::byte_order::network};
+      return broker::subnet(addr, x.length());
+    },
+    [](vast::port x) -> broker::data {
+      // We rely on the fact that port types don't change...ever.
+      auto protocol = static_cast<broker::port::protocol>(x.type());
+      return broker::port{x.number(), protocol};
+    },
+    [](vast::enumeration x) -> broker::data {
+      return broker::count{x};
```

Comment:
Did you mean `broker::enum_value`?

Reply:
That just wraps a `std::string`. In VAST, I was thinking of an enum type as a numeric constant where the type information contains the names.

I'm not sure what the best way is to consolidate these different notions. I certainly warrants a clear comment. Do you have a suggestion?

Reply:
Does this mean we can't convert safely from `broker::enum_value` to `vast::enumeration`? It seems like this approach would essentially type-erase `vast::enumeration`, i.e., when I convert a `vast::data` with an enumeration inside to `broker::data` and back it will have the type `count`.

If the types aren't 100% compatible than we can't do much about it, but at least should clearly document this (in code and user docs).

Reply:
> Does this mean we can't convert safely from broker::enum_value to vast::enumeration?

Right, not at this point. Since VAST enums are pretty much not used, I added a FIXME note for now. We can revisit this when VAST actually maps input to an enumeration.